### PR TITLE
T24867

### PR DESCRIPTION
**Pull request by dnohales**
I know this is kind of a weird approach, please let me know what do you think.

I actually couldn't test g-c-c with the proper vendor file with backgrounds directory pointing to `/var/lib/flatpak/app/com.endlessm.HackComponents/...` since I had problems on installing a locally built `com.endlessm.HackComponents` as a Flatpak, gonna keep trying.

But replacing `@BACKGROUNDSDIR@` with a hardcoded directory with the images works on g-c-c.

https://phabricator.endlessm.com/T24867

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+configure_file(
+    command: ['subst-script', '@INPUT@', '@OUTPUT@'],
+    input: 'flatpak-overrides.in',
+    output: 'global',
+    install_dir: join_paths(get_option('datadir'), 'flatpak-overrides')
+)
+
+meson.add_install_script(
```
Comment (cosimoc) [inline meson.build:32]: I wonder if `run_command()` wouldn't be a more natural way to do this.

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+configure_file(
+    input: 'hack-backgrounds/hack-backgrounds.xml.in',
+    output: 'hack-backgrounds.xml',
+    install_dir: join_paths(get_option('datadir'), 'gnome-background-properties'),
+    configuration: conf
+)
+
+configure_file(
+    command: ['subst-script', '@INPUT@', '@OUTPUT@'],
```
Comment (cosimoc) [inline meson.build]: Doesn't meson have a native way to do this? You can create a configuration object that has `FLATPAK_ARCH` and `BRANCH` set.

File: meson.build [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+install_subdir(
+    'hack-backgrounds/backgrounds',
+    install_dir: 'hack-backgrounds',
+    strip_directory: true
+)
+
+cmd = run_command('sh', '-c', 'echo $BRANCH')
+branch = cmd.stdout().strip()
```
Comment (cosimoc) [inline meson.build]: Now that we have meson, this (and I think the one below as well) could be passed as configuration options, no?

File: meson.build [hunk 4]
```diff
@@ -0,0 +1,46 @@
+project('hack-components', version: '0.0.0')
+
+install_data(
+    'hack-data.ini',
+    install_dir: join_paths(get_option('datadir'), meson.project_name())
+)
+
+install_subdir(
+    'hack-backgrounds/backgrounds',
+    install_dir: 'hack-backgrounds',
```
Comment (cosimoc) [inline meson.build]: I'd probably put the backgrounds inside the datadir as well.

**Review by cosimoc — CHANGES_REQUESTED** [review]

File: meson.build (hunk 2 above)
Comment (dnohales) [reply meson.build]: Ah yes, I forgot I was able to extract those variables. I think I could easily remove the `subst-script` file.

File: meson.build (hunk 1 above)
Comment (dnohales) [reply meson.build:32]: `run_command` runs the commands during the configuration step (where I don't think the prefix dir even exists), and this is necessarily an install step.

**Comment by dnohales** [comment]
Updated

File: meson.build (hunk 1 above)
Comment (cosimoc) [reply meson.build:32]: Makes sense.

File: meson.build [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+cmd = run_command('sh', '-c', 'echo $BRANCH')
+conf.set('BRANCH', cmd.stdout().strip())
+
+backgrounds_install_dir = join_paths(get_option('datadir'), 'backgrounds')
+conf.set('BACKGROUNDSDIR', join_paths(
+    '/var/lib/flatpak/app/com.endlessm.HackComponents',
+    conf.get('FLATPAK_ARCH'),
+    conf.get('BRANCH'),
+    'active/files',
+    backgrounds_install_dir
```
Comment (cosimoc) [inline meson.build]: Are you sure that this does not include the prefix? Otherwise we'll set a path that does not exist.

File: meson.build [hunk 6]
```diff
@@ -0,0 +1,49 @@
+project('hack-components', version: '0.0.0')
+
+conf = configuration_data()
+
+cmd = run_command('sh', '-c', 'echo $FLATPAK_ARCH')
+conf.set('FLATPAK_ARCH', cmd.stdout().strip())
+
+cmd = run_command('sh', '-c', 'echo $BRANCH')
+conf.set('BRANCH', cmd.stdout().strip())
```
Comment (cosimoc) [inline meson.build]: I still would like to get these environment variables out of the way if possible... At least for `$BRANCH` we can pass it as an actual meson configuration option, no?

File: meson.build [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+backgrounds_install_dir = join_paths(get_option('datadir'), 'backgrounds')
+conf.set('BACKGROUNDSDIR', join_paths(
+    '/var/lib/flatpak/app/com.endlessm.HackComponents',
```
Comment (cosimoc) [inline meson.build]: Something that occurred to me is that this won't work if HackComponents is installed in the user repository.

However we could easily make the control center support relative paths from the location of the XML file, which would make thins things a lot simpler here. You can probably just change this [1] to use `g_file_new_for_commandline_arg_and_cwd()` and pass the directory of the XML file as the cwd.

[1] https://github.com/endlessm/gnome-control-center/blob/17c90fdcb4ad62a1ef0964ee439c5b1f553a2388/panels/background/cc-background-xml.c#L209

**Review by cosimoc — CHANGES_REQUESTED** [review]

File: meson.build (hunk 5 above)
Comment (dnohales) [reply meson.build]: I have tested it:

```
$ tree build
build
├── export
├── files
│   ├── clippy
│   ├── hack-avatar-faces
│   ├── manifest.json
│   └── share
│       ├── backgrounds
│       │   ├── Desktop-BGs-Beta-Binary.png
│       │   ├── Desktop-BGs-Beta-Cat_Grid.png
│       │   ├── Desktop-BGs-Beta-Pop_Art.png
│       │   ├── Desktop-BGs-Beta-Sketch_Blue.png
│       │   ├── Desktop-BGs-Beta-Sketch_Green.png
│       │   ├── hack_colors_05_blurred-FBE.png
│       │   ├── hack_colors_05.png
│       │   ├── riley_colors_02b_blurred-FBE.png
│       │   ├── riley_colors_02b.png
│       │   └── saniel_colors_02b.png
│       ├── flatpak-overrides
│       │   └── global
│       ├── gnome-background-properties
│       │   └── hack-backgrounds.xml
│       └── hack-components
│           └── hack-data.ini
├── metadata
└── var
    ├── lib
    ├── run -> /run
    └── tmp
```

```
$ head build/files/share/gnome-background-properties/hack-backgrounds.xml 
<?xml version="1.0"?>
<!DOCTYPE wallpapers SYSTEM "gnome-wp-list.dtd">
<wallpapers>
  <wallpaper deleted="false">
    <name>Desktop-BGs-Beta-Binary.png</name>
    <filename>/var/lib/flatpak/app/com.endlessm.HackComponents/x86_64/master/active/files/share/backgrounds/Desktop-BGs-Beta-Binary.png</filename>
    <options>zoom</options>
    <shade_type>solid</shade_type>
    <pcolor>#3465a4</pcolor>
    <scolor>#000000</scolor>
```

File: meson.build (hunk 6 above)
Comment (dnohales) [reply meson.build]: Yes, we could.

**Pull request merged**